Firestore/core/src/firebase/firestore/model/document_set.h

```diff
@@ -0,0 +1,178 @@
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_DOCUMENT_SET_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_MODEL_DOCUMENT_SET_H_
+
+#import <Foundation/Foundation.h>
+
+#include <iosfwd>
+#include <string>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/immutable/sorted_set.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_map.h"
+
+@class FSTDocument;
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+/**
+ * A C++ comparator that returns less-than, implemented by delegating to
+ * an NSComparator.
+ */
+class DocumentSetComparator {
+ public:
+  explicit DocumentSetComparator(NSComparator delegate = nil)
+      : delegate_(delegate) {
+  }
+
+  bool operator()(FSTDocument* lhs, FSTDocument* rhs) const {
+    return delegate_(lhs, rhs) == NSOrderedAscending;
+  }
+
+ private:
+  NSComparator delegate_;
+};
+
+/**
+ * DocumentSet is an immutable (copy-on-write) collection that holds documents
+ * in order specified by the provided comparator. We always add a document key
+ * comparator on top of what is provided to guarantee document equality based on
+ * the key.
+ */
+class DocumentSet {
+ public:
+  /**
+   * The type of the main collection of documents in an DocumentSet.
+   * @see sorted_set_.
+   */
+  using SetType = immutable::SortedSet<FSTDocument*, DocumentSetComparator>;
+
+  // STL container types
+  using value_type = FSTDocument*;
+  using const_iterator = SetType::const_iterator;
+
+  /**
+   * Creates a new, empty DocumentSet sorted by the given comparator, then by
+   * keys.
+   */
+  explicit DocumentSet(NSComparator comparator);
+
+  size_t size() const {
+    return index_.size();
+  }
+
+  /** Returns true if the dictionary contains no elements. */
+  bool empty() const {
+    return index_.empty();
+  }
+
+  /** Returns true if this set contains a document with the given key. */
+  bool ContainsKey(const DocumentKey& key) const;
+
+  SetType::const_iterator begin() const {
+    return sorted_set_.begin();
+  }
+  SetType::const_iterator end() const {
+    return sorted_set_.end();
+  }
+
+  /**
+   * Returns the document from this set with the given key if it exists or nil
+   * if it doesn't.
+   */
+  FSTDocument* _Nullable GetDocument(const DocumentKey& key) const;
+
+  /**
+   * Returns the first document in the set according to its built in ordering,
+   * or nil if the set is empty.
+   */
+  FSTDocument* _Nullable GetFirstDocument() const;
+
+  /**
+   * Returns the last document in the set according to its built in ordering, or
+   * nil if the set is empty.
+   */
+  FSTDocument* _Nullable GetLastDocument() const;
+
+  /**
+   * Returns the index of the document with the provided key in the document
+   * set. Returns NSNotFound if the key is not present.
```

Comment:
Given your followup patch, I think this should say "Returns npos if the key is not present"